internal/agentcfg/fetch.go

```diff
@@ -218,5 +91,84 @@ func (f *DirectFetcher) Fetch(_ context.Context, query Query) (Result, error) {
 		}}
 	}
 
-	return sanitize(query.InsecureAgents, result), nil
+	return result, nil
+}
+
+func matchAgentConfig(query Query, cfgs []AgentConfig) Result {
+	name, env := query.Service.Name, query.Service.Environment
+	result := zeroResult()
+	var nameConf, envConf, defaultConf *AgentConfig
+
+	for i, cfg := range cfgs {
+		if cfg.ServiceName == name && cfg.ServiceEnvironment == env {
+			nameConf = &cfgs[i]
+			break
+		} else if cfg.ServiceName == name && cfg.ServiceEnvironment == "" {
+			nameConf = &cfgs[i]
+		} else if cfg.ServiceName == "" && cfg.ServiceEnvironment == env {
+			envConf = &cfgs[i]
+		} else if cfg.ServiceName == "" && cfg.ServiceEnvironment == "" {
+			defaultConf = &cfgs[i]
+		}
+	}
+
+	if nameConf != nil {
+		result = Result{Source{
+			Settings: nameConf.Config,
+			Etag:     nameConf.Etag,
+			Agent:    nameConf.AgentName,
+		}}
+	} else if envConf != nil {
+		result = Result{Source{
+			Settings: envConf.Config,
+			Etag:     envConf.Etag,
+			Agent:    envConf.AgentName,
+		}}
+	} else if defaultConf != nil {
+		result = Result{Source{
+			Settings: defaultConf.Config,
+			Etag:     defaultConf.Etag,
+			Agent:    defaultConf.AgentName,
+		}}
+	}
```

Comment:
This is a duplication of https://github.com/elastic/apm-server/pull/9720/files#diff-4c475b639581f99589dafac34d999b297fa831132b0db140c99a029ef3918282R61-R92 - please avoid copying the exact logic and reuse instead.

Reply:
sorry, bad manual rebase :facepalm: